### PR TITLE
feat(storage): track GC watermark SST id.

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR tracks GC watermark SST id in compute node and compactor node, in order to filter out false positive orphan SSTs from object store during full SST GC.

- An orphan SST is one that exists in object store, but not found in hummock version.
- A false positive orphan SST is one that belongs to an ongoing write op. It may be added to hummock version later. So it's fatal to delete it during full SST GC.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/4037

### CODE REVIEW

File: src/storage/src/hummock/compactor.rs [hunk 1]
```diff
@@ -449,10 +449,30 @@ impl Compactor {
...
+        // write op. The watermark is invalidated when this method exits.
+        let tracker_id = match context.sstable_id_manager.add_watermark_sst_id(None).await {
+            Ok(tracker_id) => tracker_id,
+            Err(err) => {
+                tracing::warn!("Failed to track pending SST id. {:#?}", err);
+                return false;
+            }
+        };
+        let sstable_id_manager_clone = context.sstable_id_manager.clone();
+        scopeguard::guard(
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor.rs:467]: Dumb question: if we don't assign a variable to this statement, is it possible for the rust compiler to drop the scopeguard before this function ends?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM.

File: src/storage/src/hummock/shared_buffer/shared_buffer_uploader.rs [hunk 2]
```diff
@@ -114,6 +114,14 @@ impl SharedBufferUploader {
             self.remote_object_store_compactor_context.clone()
         };
 
+        // Set a watermark SST id for this epoch to prevent full GC from accidentally deleting SSTs
+        // for in-progress write op. The watermark is invalidated when the epoch is
+        // committed or cancelled.
+        mem_compactor_ctx
+            .sstable_id_manager
+            .add_watermark_sst_id(Some(epoch))
+            .await?;
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/src/hummock/shared_buffer/shared_buffer_uploader.rs:123]: We can merge this PR first and modify it when barrier-checkpoint decoupling is ready.

File: src/storage/src/hummock/compactor.rs (hunk 1 above)
Comment (zwang28) [reply src/storage/src/hummock/compactor.rs:467]: You're right. Thx

**Pull request merged**